### ISSUE TITLE
Mapped signal leaks if it doesn't terminate

### ISSUE THREAD

**Issue by Vadim-Yelagin** [ReactiveCocoa/ReactiveCocoa#2586]
This code causes a genuine memory leak

	let (signal, _) = Signal<Int, NoError>.pipe()
	let _ = signal.map { $0 }

<img width="985" alt="screenshot 2015-12-02 16 27 30" src="https://cloud.githubusercontent.com/assets/5680096/11528564/61f16856-9909-11e5-8439-fc51035f9789.png">

**Comment by Vadim-Yelagin** [comment]
Looking at implementation of `init`, `observe` and `map` I see that the retain cycle:
(original signal) -> (original signal observers) -> (observer added by map) -> (mapped signal's internal observer) -> (mapped signal's generator disposable) -> (disposable from original signal's observe in map) -> (original signal)

**Comment by NachoSoto** [comment]
Cool I guess that confirms the bug!
Let's keep the conversation in #2585 to avoid duplication.

**Comment by Vadim-Yelagin** [comment]
There might be a way to fix this without affecting https://github.com/ReactiveCocoa/ReactiveCocoa/issues/2585 or breaking the signal contract. There is nothing that says that a signal must be retained by its observers or corresponding disposables. The only thing that must retain the signal is its generator observable. While in https://github.com/ReactiveCocoa/ReactiveCocoa/issues/2585 the original generator is not deallocated and the signal might potentially generate events and even terminate (which is the only thing that can break the retain cycle), in case of `Signal.never` the generator does not retain the observer it's provided with.
What I propose is to prevent the retain cycle here by making observing disposable capture the signal weakly. It will allow the original signal (e.g. `Signal.never`) to get deallocated, even though it has observers, since it cannot generate any events anymore.

**Comment by Vadim-Yelagin** [comment]
TL;DR Might be not duplicate, please see https://github.com/ReactiveCocoa/ReactiveCocoa/pull/2603
